torch_geometric/utils/dropout.py

```diff
@@ -208,3 +214,94 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
+    r"""Drops edges from the adjacency matrix :obj:`edge_index`
+    based on random walks. The source nodes to start random walks from are
+    sampled with probability :obj:`r` from a Bernoulli distribution.
+
+    The method returns (1) the retained :obj:`edge_index`, (2) the edge mask
+    or index indicating which edges were retained, depending on the argument
+    :obj:`force_undirected`.
+
+    Args:
+        edge_index (LongTensor): The edge indices.
+        r (float, optional): The ratio of source nodes to start
+            random walks from. (default: :obj:`0.5`)
+        force_undirected (bool, optional): If set to :obj:`True`, will either
+            drop or keep both edges of an undirected edge.
+            (default: :obj:`False`)
+        walks_per_node (int, optional): The number of walks per node.
+            (default: :obj:`1`)
+        walk_length (int, optional): The length per walk. (default: :obj:`3`)
+        p (float, optional): :obj:`p` in random walks. (default: :obj:`1`)
+        q (float, optional): :obj:`q` in random walks. (default: :obj:`1`)
+        num_nodes (int, optional): The number of nodes, *i.e.*
+            :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)
+        training (bool, optional): If set to :obj:`False`, this operation is a
+            no-op. (default: :obj:`True`)
+
+    :rtype: (:class:`LongTensor`, :class:`BoolTensor` or :class:`LongTensor`)
+
+    Examples:
+
+        >>> edge_index = torch.tensor([[0, 1, 1, 2, 2, 3],
+        ...                            [1, 0, 2, 1, 3, 2]])
+        >>> edge_index, edge_mask = dropout_path(edge_index)
+        >>> edge_index
+        tensor([[1, 2],
+                [2, 3]])
+        >>> edge_mask # masks indicating which edges are retained
+        tensor([False, False,  True, False,  True, False])
+
+        >>> edge_index, edge_id = dropout_path(edge_index,
+        ...                                    force_undirected=True)
+        >>> edge_index
+        tensor([[2, 3],
+                [3, 2]])
+        >>> edge_id # indices indicating which edges are retained
+        tensor([4, 4])
+    """
+
+    if r < 0. or r > 1.:
+        raise ValueError(f'Sampling ratio has to be between 0 and 1 '
+                         f'(got {r}')
+
+    edge_mask = edge_index.new_ones(edge_index.size(1), dtype=torch.bool)
+    if not training or r == 0.0:
+        return edge_index, edge_mask
+
+    if random_walk is None:
+        raise ImportError('`dropout_path` requires `torch-cluster`.')
+
+    row, col = edge_index
+    num_nodes = maybe_num_nodes(edge_index, num_nodes)
+    deg = torch.zeros(num_nodes, device=row.device)
+    deg.scatter_add_(0, row, torch.ones(row.size(0), device=row.device))
+
+    num_starts = round(r * num_nodes)
+    start = torch.randperm(num_nodes, device=edge_index.device)[:num_starts]
+    start = start.repeat(walks_per_node)
+
+    rowptr = row.new_zeros(num_nodes + 1)
```

Comment:
This does not look correct to me. We would first need to sort `edge_index` row-wise. Otherwise, `col` does not align with `rowptr`. Note that in this case, you also need to permute the resulting `edge_id`.

Reply:
Thanks! It has been fixed with an `is_sorted` argument.

Reply:
Not yet. Also need to permute the resulting `edge_id`.

Reply:
Done. Argument `force_undirected` has been dropped to make it simple.